File: scripts/batch-enrich-jgrants.mjs
/**
 * JGrants バッチエンリッチメントスクリプト
 * 
 * V2 APIから詳細情報を取得し、本番DBを更新する
 * 
 * 使用方法:
 *   node scripts/batch-enrich-jgrants.mjs
 */

import { execSync } from 'child_process';

const JGRANTS_V2_API = 'https://api.jgrants-portal.go.jp/exp/v2/public/subsidies/id';

// 対象IDリスト
const TARGET_IDS = [
  'a0WJ200000CDQEJMA5',
  'a0WJ200000CDX4vMAH',
  'a0WJ200000CDVkeMAH',
  'a0WJ200000CDPyeMAH',
  'a0W2x000007Cos5EAC',
  'a0WJ200000CDTzMMAX',
  'a0WJ200000CDPbMMAX',
  'a0WJ200000CDWemMAH',
  'a0WJ200000CDUGBMA5',
  'a0WJ200000CDUg2MAH',
];

/**
 * HTML からテキストを抽出
 */
function htmlToText(html, maxLength = 3000) {
  if (!html) return '';
  return html
    .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
    .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/\s+/g, ' ')
    .trim()
    .substring(0, maxLength);
}

/**
 * HTML から PDF リンクを抽出
 */
function extractPdfLinks(html) {
  if (!html) return [];
  const urls = [];
  const pattern = /https?:\/\/[^\s"'<>]+\.pdf/gi;
  let match;
  while ((match = pattern.exec(html)) !== null) {
    if (!urls.includes(match[0])) {
      urls.push(match[0]);
    }
  }
  return urls;
}

/**
 * V2 API から詳細情報を取得
 */
async function fetchSubsidyDetail(id) {
  const response = await fetch(`${JGRANTS_V2_API}/${id}`, {
    headers: { 'Accept': 'application/json' },
  });
  
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  
  const data = await response.json();
  const subsidy = data.result?.[0] || data.result || data;
  
  if (!subsidy) {
    throw new Error('No result in response');
  }
  
  return subsidy;
}

/**
 * detail_json を構築
 */
function buildDetailJson(subsidy) {
  const detailJson = {
    enriched_at: new Date().toISOString(),
    enriched_version: 'v2',
    api_version: 'v2',
  };
  
  // Overview
  if (subsidy.detail) {
    detailJson.overview = htmlToText(subsidy.detail, 3000);
  }
  
  // Basic info
  if (subsidy.subsidy_catch_phrase) {
    detailJson.catch_phrase = subsidy.subsidy_catch_phrase;
  }
  if (subsidy.use_purpose) {
    detailJson.use_purpose = subsidy.use_purpose;
  }
  if (subsidy.industry) {
    detailJson.target_industry = subsidy.industry;
  }
  if (subsidy.target_number_of_employees) {
    detailJson.target_employees = subsidy.target_number_of_employees;
  }
  if (subsidy.subsidy_rate) {
    detailJson.subsidy_rate = subsidy.subsidy_rate;
  }
  if (subsidy.subsidy_max_limit) {
    detailJson.subsidy_max_limit = subsidy.subsidy_max_limit;
  }
  if (subsidy.granttype) {
    detailJson.grant_type = subsidy.granttype;
  }
  if (subsidy.front_subsidy_detail_page_url) {
    detailJson.related_url = subsidy.front_subsidy_detail_page_url;
  }
  
  // Workflows
  if (subsidy.workflow && Array.isArray(subsidy.workflow) && subsidy.workflow.length > 0) {
    detailJson.workflows = subsidy.workflow.map(w => ({
      id: w.id,
      target_area: w.target_area_search,
      acceptance_start: w.acceptance_start_datetime,
      acceptance_end: w.acceptance_end_datetime,
      project_end: w.project_end_deadline,
    }));
  }
  
  // PDF URLs
  if (subsidy.detail) {
    const pdfUrls = extractPdfLinks(subsidy.detail);
    if (pdfUrls.length > 0) {
      detailJson.pdf_urls = pdfUrls;
    }
  }
  
  return detailJson;
}

// Note: DB更新はSQL出力のみ（手動実行用）

/**
 * メイン処理
 */
async function main() {
  console.log('='.repeat(60));
  console.log('JGrants Batch Enrichment Script (V2 API)');
  console.log('='.repeat(60));
  console.log(`Targets: ${TARGET_IDS.length} subsidies`);
  console.log('');
  
  let successCount = 0;
  let errorCount = 0;
  const results = [];
  
  for (const id of TARGET_IDS) {
    console.log(`Processing: ${id}`);
    
    try {
      // V2 API から取得
      const subsidy = await fetchSubsidyDetail(id);
      console.log(`  Title: ${subsidy.title?.substring(0, 50)}...`);
      
      // detail_json 構築
      const detailJson = buildDetailJson(subsidy);
      const pdfCount = detailJson.pdf_urls?.length || 0;
      console.log(`  Detail length: ${detailJson.overview?.length || 0} chars`);
      console.log(`  PDF URLs: ${pdfCount}`);
      
      // 結果を保存（後で一括更新）
      results.push({ id, detailJson, title: subsidy.title });
      successCount++;
      
      // レート制限
      await new Promise(resolve => setTimeout(resolve, 500));
      
    } catch (err) {
      console.log(`  ERROR: ${err.message}`);
      errorCount++;
    }
  }
  
  console.log('');
  console.log('-'.repeat(60));
  console.log(`Fetched: ${successCount} success, ${errorCount} errors`);
  console.log('');
  
  // 結果を出力
  console.log('Results (copy to execute manually):');
  console.log('');
  
  for (const { id, detailJson } of results) {
    const jsonStr = JSON.stringify(detailJson).replace(/'/g, "''");
    console.log(`-- ${id}`);
    console.log(`UPDATE subsidy_cache SET detail_json = '${jsonStr}', cached_at = datetime('now') WHERE id = '${id}';`);
    console.log('');
  }
}

main().catch(console.error);

File: tools/batch-enrich-jgrants.py
#!/usr/bin/env python3
"""
jGrants受付中の未enrich件を一括enrich → 本番DBに書き込む

Usage:
  python3 tools/batch-enrich-jgrants.py
"""

import json
import subprocess
import time
import re
import sys
import urllib.request

WRANGLER_ENV = {
    "CLOUDFLARE_API_KEY": "f20c372dc7eff0b088b3f1e4b3ec03592bda6",
    "CLOUDFLARE_EMAIL": "sekiyadubai@gmail.com",
}
DB_NAME = "subsidy-matching-production"
JGRANTS_V2_API = "https://api.jgrants-portal.go.jp/exp/v2/public/subsidies/id"

def run_d1(sql, raw=False):
    """wrangler d1 executeで本番DBにSQLを実行（ファイル経由で長いSQL対応）"""
    import os
    import tempfile
    env = os.environ.copy()
    env.update(WRANGLER_ENV)
    
    # 長いSQLはファイル経由で実行
    if len(sql) > 5000:
        with tempfile.NamedTemporaryFile(mode='w', suffix='.sql', delete=False, dir='/tmp') as f:
            f.write(sql)
            sql_file = f.name
        cmd = [
            "npx", "wrangler", "d1", "execute", DB_NAME,
            "--remote", "--file", sql_file
        ]
    else:
        cmd = [
            "npx", "wrangler", "d1", "execute", DB_NAME,
            "--remote", "--command", sql
        ]
    
    result = subprocess.run(cmd, capture_output=True, text=True, env=env, cwd="/home/user/webapp")
    
    # 一時ファイルのクリーンアップ
    if len(sql) > 5000:
        try:
            os.unlink(sql_file)
        except:
            pass
    if raw:
        return result.stdout
    # JSONからresultsを抽出
    try:
        # wranglerの出力からJSON部分を抽出
        output = result.stdout
        # "results" : [...] を探す
        match = re.search(r'"results"\s*:\s*\[', output)
        if match:
            start = match.start()
            # 対応する ] を見つける
            bracket_count = 0
            for i, c in enumerate(output[match.end()-1:], match.end()-1):
                if c == '[':
                    bracket_count += 1
                elif c == ']':
                    bracket_count -= 1
                    if bracket_count == 0:
                        json_str = '{' + output[start:i+1] + '}'
                        return json.loads(json_str)['results']
        return []
    except Exception as e:
        print(f"Parse error: {e}")
        print(f"Stdout: {result.stdout[:500]}")
        return []

def fetch_jgrants_detail(subsidy_id):
    """jGrants v2 APIから詳細情報を取得"""
    url = f"{JGRANTS_V2_API}/{subsidy_id}"
    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode())
            result = data.get("result", [])
            if isinstance(result, list) and result:
                return result[0]
            return result if result else None
    except Exception as e:
        print(f"  API error for {subsidy_id}: {e}")
        return None

def build_detail_json(subsidy, existing_detail_json=None):
    """APIレスポンスからdetail_jsonを構築"""
    existing = json.loads(existing_detail_json) if existing_detail_json else {}
    
    # HTML tagを除去
    detail_text = subsidy.get("detail", "") or ""
    overview = re.sub(r'<[^>]*>', ' ', detail_text).strip()
    overview = re.sub(r'\s+', ' ', overview)[:3000]
    
    # PDF URLを抽出
    pdf_urls = []
    pdf_matches = re.findall(r'https?://[^\s"\'<>]+\.pdf', detail_text, re.IGNORECASE)
    for url in pdf_matches:
        if url not in pdf_urls:
            pdf_urls.append(url)
    
    # workflowを整理
    workflows = []
    for w in (subsidy.get("workflow") or []):
        workflows.append({
            "id": w.get("id"),
            "target_area": w.get("target_area"),
            "acceptance_start": w.get("acceptance_start_datetime"),
            "acceptance_end": w.get("acceptance_end_datetime"),
            "project_end": w.get("project_end_deadline"),
        })
    
    detail_json = {
        **existing,  # 既存データを保持
        "enriched_at": time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime()),
        "enriched_version": "v2",
        "api_version": "v2",
        "overview": overview if overview else existing.get("overview"),
        "catch_phrase": subsidy.get("subsidy_catch_phrase") or existing.get("catch_phrase"),
        "use_purpose": subsidy.get("use_purpose") or existing.get("use_purpose"),
        "target_industry": (
            " / ".join(subsidy["industry"]) if isinstance(subsidy.get("industry"), list)
            else subsidy.get("industry") or existing.get("target_industry")
        ),
        "target_employees": subsidy.get("target_number_of_employees") or existing.get("target_employees"),
        "subsidy_rate": subsidy.get("subsidy_rate") or existing.get("subsidy_rate"),
        "subsidy_max_limit": subsidy.get("subsidy_max_limit") or existing.get("subsidy_max_limit"),
        "grant_type": subsidy.get("granttype") or existing.get("grant_type"),
        "related_url": f"https://www.jgrants-portal.go.jp/subsidy/{subsidy.get('id', '')}",
        "workflows": workflows if workflows else existing.get("workflows", []),
    }
    
    if pdf_urls:
        detail_json["pdf_urls"] = pdf_urls
    
    # application_guidelines があれば追加
    guidelines = subsidy.get("application_guidelines") or []
    if guidelines:
        detail_json["application_guidelines"] = guidelines
    
    return detail_json

def main():
    print("=== jGrants受付中 未enrich件 一括enrich ===")
    print()
    
    # 1. 未enrich件のIDリストを取得
    print("[1] 未enrich件を取得中...")
    rows = run_d1("""
        SELECT id, title, detail_json FROM subsidy_cache 
        WHERE source='jgrants' 
          AND (acceptance_end_datetime > datetime('now') OR acceptance_end_datetime IS NULL)
          AND (json_extract(detail_json,'$.enriched_version') IS NULL OR json_extract(detail_json,'$.enriched_version') != 'v2')
        ORDER BY acceptance_end_datetime ASC
    """)
    
    print(f"  対象: {len(rows)}件")
    if not rows:
        print("  全件enrich済み!")
        return
    
    # 2. 各件をenrich
    enriched = 0
    errors = 0
    ready_candidates = 0
    
    for i, row in enumerate(rows):
        sid = row["id"]
        title = (row.get("title") or "")[:50]
        print(f"\n[{i+1}/{len(rows)}] {sid} - {title}...")
        
        # jGrants v2 APIから取得
        detail = fetch_jgrants_detail(sid)
        if not detail:
            errors += 1
            continue
        
        # detail_jsonを構築
        detail_json = build_detail_json(detail, row.get("detail_json"))
        detail_json_str = json.dumps(detail_json, ensure_ascii=False)
        
        # SQLエスケープ（シングルクォートを2つに）
        escaped = detail_json_str.replace("'", "''")
        
        # 本番DBに書き込み
        update_sql = f"UPDATE subsidy_cache SET detail_json = '{escaped}', cached_at = datetime('now') WHERE id = '{sid}'"
        run_d1(update_sql, raw=True)
        
        enriched += 1
        has_overview = bool(detail_json.get("overview"))
        print(f"  -> enriched (overview={len(detail_json.get('overview',''))}chars, pdfs={len(detail_json.get('pdf_urls',[]))})")
        
        # Rate limiting
        time.sleep(0.5)
    
    print(f"\n=== 完了 ===")
    print(f"  対象: {len(rows)}件")
    print(f"  enriched: {enriched}件")
    print(f"  errors: {errors}件")
    
    # 3. daily-ready-boost を実行して ready 化
    print(f"\n[3] daily-ready-boost を実行中...")
    result = subprocess.run(
        ["curl", "-s", "-X", "POST", 
         "https://hojyokin-cron.sekiyadubai.workers.dev/trigger",
         "-H", "Content-Type: application/json",
         "-d", '{"job":"sync"}'],
        capture_output=True, text=True, timeout=120
    )
    print(f"  -> {result.stdout[:200]}")

if __name__ == "__main__":
    main()
